### PR TITLE
report: better deduping of warnings

### PR DESCRIPTION
**Pull request by jckr**
**Summary**
This is in response to issue #11514. Currently, gather-runner can dedupe warnings if they are strings, however, they can also be of type LH.IcuMessage, and as a result several such warnings can appear in the final report. 

In my previous PR #12351, that problem could appear:

image

**Related Issues/PRs**

### CODE REVIEW

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -626,7 +626,18 @@ class GatherRunner {
    */
   static finalizeBaseArtifacts(baseArtifacts) {
     // Take only unique LighthouseRunWarnings.
-    baseArtifacts.LighthouseRunWarnings = Array.from(new Set(baseArtifacts.LighthouseRunWarnings));
+
+    const uniqueWarnings = new Set();
+    const LighthouseRunWarnings = [];
+    for (const warning of baseArtifacts.LighthouseRunWarnings) {
+      // each entry in baseArtifacts.LighthouseRunWarnings can be a string or a LH.IcuMessage.
+      const stringifiedWarning = JSON.stringify(warning);
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js:634]: We have the `lodash.isequal` package in our dependencies which does deep equality like what we're looking for here and isn't sensitive to ordering like a stringify approach is. 

Given we tend to have very few warnings here, I'm not as concerned with preserving hashset time complexity :)

WDYT about switching to a `LighthouseRunWarnings.some(existing => isDeepEqual(warning, existing)` check?

**Review by patrickhulce — COMMENTED** [review]
nice work! thanks @jckr :)

File: lighthouse-core/gather/gather-runner.js (hunk 1 above)
Comment (jckr) [reply lighthouse-core/gather/gather-runner.js]: i see what you mean. updating the branch

**Review by patrickhulce — APPROVED** [review]
LGTM 💯 

great stuff!

**Comment by paulirish** [comment]
looks like this is good to go. 

the type change seems optional. i'm fine with either way.

**Comment by paulirish** [comment]
Heh, looks like @patrickhulce actually lifted this implementation when he landed https://github.com/GoogleChrome/lighthouse/commit/c5fddd9e5520fea397ed498baba52b69f676bbb0#diff-19b3c3e228a1602f508fa6e6df814df9d1e45475f96894455df2f6eb08bf67baR85

However the test wasn't modified.  So I've made this PR current and now it's just the test. \o/

**Comment by paulirish** [comment]
@googlebot I consent.

**Pull request merged**